ip_proofs/src/gipa.rs

```diff
@@ -0,0 +1,556 @@
+        let (com_a, com_b, com_t) = com;
+        match ck_a.len() {
+            1 => {
+                let a_base = vec![proof.r_base.0.clone()];
+                let b_base = vec![proof.r_base.1.clone()];
+                let t_base = vec![IP::inner_product(&a_base, &b_base)?];
+                Ok(LMC::verify(ck_a, &a_base, com_a)?
+                    && RMC::verify(ck_b, &b_base, com_b)?
+                    && IPC::verify(ck_t, &t_base, com_t)?)
+            } // base case
+            2..=usize::MAX if ck_a.len().count_ones() == 1 => {
+                // recursive step
+                // Fiat-Shamir challenge
+                let (com_1, com_2) = proof.r_commitment_steps.pop().unwrap();
+                let mut counter_nonce: usize = 0;
+                let (c, c_inv) = loop {
+                    let mut hash_input = Vec::new();
+                    hash_input.extend_from_slice(&counter_nonce.to_be_bytes()[..]);
+                    hash_input.extend_from_slice(&to_bytes![
+                        transcript, com_1.0, com_1.1, com_1.2, com_2.0, com_2.1, com_2.2
+                    ]?);
+                    if let Some(c) = LMC::Scalar::from_random_bytes(&D::digest(&hash_input)) {
+                        if let Some(c_inv) = c.inverse() {
+                            break (c, c_inv);
+                        }
+                    };
+                    counter_nonce += 1;
+                };
+
+                let split = ck_a.len() / 2;
+                let ck_a_1 = &ck_a[..split];
+                let ck_a_2 = &ck_a[split..];
+                let ck_b_1 = &ck_b[split..];
+                let ck_b_2 = &ck_b[..split];
+
+                let ck_a_recurse = ck_a_2
+                    .iter()
+                    .map(|a| mul_helper(a, &c_inv))
+                    .zip(ck_a_1)
+                    .map(|(a_1, a_2)| a_1.clone() + a_2.clone())
+                    .collect::<Vec<LMC::Key>>();
+
+                let ck_b_recurse = ck_b_1
+                    .iter()
+                    .map(|b| mul_helper(b, &c))
+                    .zip(ck_b_2)
+                    .map(|(b_1, b_2)| b_1.clone() + b_2.clone())
+                    .collect::<Vec<RMC::Key>>();
+
+                let com_recurse = (
+                    mul_helper(&com_1.0, &c) + com_a.clone() + mul_helper(&com_2.0, &c_inv),
+                    mul_helper(&com_1.1, &c) + com_b.clone() + mul_helper(&com_2.1, &c_inv),
+                    mul_helper(&com_1.2, &c) + com_t.clone() + mul_helper(&com_2.2, &c_inv),
+                );
+
+                Self::recursive_verify(
+                    (&ck_a_recurse, &ck_b_recurse, ck_t),
+                    (&com_recurse.0, &com_recurse.1, &com_recurse.2),
+                    proof,
+                    &c,
+                )
+            }
+            _ => unreachable!(), // If called only on message lengths power of 2
+        }
+    }
+}
+
+impl<IP, LMC, RMC, IPC, D> Clone for GIPAProof<IP, LMC, RMC, IPC, D>
```

Comment:
If moving to the `GIPAConfig` idea, you can just derive this.

Reply:
See other response.